invenio_sequencegenerator/api.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""User-level API."""
+
+from __future__ import absolute_import, print_function
+
+from invenio_db import db
+
+from invenio_sequencegenerator.errors import SequenceNotFound
+from invenio_sequencegenerator.models import Counter, TemplateDefinition
+
+
+class Template(object):
+    """API for defining sequences."""
+
+    def __init__(self, name, _model=None):
+        """Initialize template.
+
+        :param name: the identifying name of the template
+        :param _model: the model object of the template
+        """
+        self.model = _model or TemplateDefinition.query.get(name)
+        if self.model is None:
+            raise SequenceNotFound()
+        assert name == self.model.name
+
+    @classmethod
+    def create(cls, name, template, start=0, step=1):
+        """Create a new sequence definition.
+
+        :param name: the identifying name of the template
+        :param template: the template string of the template
+        :param start: the starting counter of sequences, based on this template
+        :param step: the incremental step
+        """
+        with db.session.begin_nested():
+            model = TemplateDefinition(
+                name=name, template=template, start=start, step=step
+            )
+            db.session.add(model)
+        return cls(name, _model=model)
+
+    def delete(self):
+        """Delete templates generated from this sequence definition."""
+        with db.session.begin_nested():
+            db.session.delete(self.model)
+
+
+class Sequence(object):
+    """Iterator for sequences."""
+
+    def __init__(self, template, **kwargs):
+        """Initialise sequence iterator, with or without keyword arguments.
+
+        :param template: the template that this sequence is based on
+        :param kwargs: the kwargs to instantiate the template
```

Comment:
We should consider the `ctx` (context) variable instead of/together with `kwargs`.